File: server/services/kundaliBilling.js
/* Kundali commercial model.

   One customer account includes a plan-wise number of free (personal) kundalis;
   every additional — and every FAMILY MEMBER — kundali is a chargeable request.
   All prices, GST, discounts, coupon eligibility and free counts come from the
   `kundali_pricing` setting (Admin > Kundali Management > Pricing). Nothing is
   hard-coded here beyond safe defaults for a fresh database.

   Billing states: FREE | PENDING_PAYMENT | PAID | REFUNDED | CANCELLED */
'use strict';
const { db, getSetting } = require('../db');
const { bad, conflict } = require('../lib/util');
const DEFAULTS = {
  active: true,
  currency: 'INR',
  personalPrice: 0,        // paid only after the plan's free quota is used up
  familyPrice: 499,        // every family-member kundali is chargeable
  additionalPrice: 499,    // extra personal kundalis beyond the quota
  gstPct: 5,
  discountPct: 0,
  couponEligible: true,
  freeCounts: { customer: 1, plus: 2, premium: 5 }
};

const pricing = () => Object.assign({}, DEFAULTS, getSetting('kundali_pricing', {}));

const includedCount = (user) => {
  const p = pricing();
  if (user.premium) return p.freeCounts.premium;
  if (user.plus) return p.freeCounts.plus;
  return p.freeCounts.customer;
};

/* Personal kundalis used against the plan quota (family kundalis never consume it). */
const usedCount = (userId) =>
  db.prepare("SELECT COUNT(*) n FROM kundalis WHERE customer_id=? AND (relationship='' OR relationship IS NULL)").get(userId).n;

/* Decide billing for a request. `relationship` empty => personal kundali. */
function classify(user, relationship) {
  const p = pricing();
  if (!p.active) throw bad('Kundali generation is currently unavailable');
  const rel = String(relationship || '').trim();
  if (rel) return { family: true, included: false, base: p.familyPrice, label: rel };
  const used = usedCount(user.id), inc = includedCount(user);
  if (used < inc) return { family: false, included: true, base: 0, label: 'Included' };
  return { family: false, included: false, base: p.additionalPrice, label: 'Additional' };
}

/* Full quote: base -> plan/special discount -> coupon -> GST. Amounts in rupees (int). */
function quoteFor(user, { relationship, coupon } = {}) {
  const p = pricing();
  const c = classify(user, relationship);
  let discount = Math.round(c.base * (p.discountPct || 0) / 100);
  let couponInfo = null;
  if (coupon && c.base > 0) {
    if (!p.couponEligible) throw bad('Coupons do not apply to kundali purchases');
    const cp = db.prepare('SELECT * FROM coupons WHERE code=? AND active=1').get(String(coupon).toUpperCase());
    if (!cp) throw bad('Coupon code is not valid');
    if (cp.min && c.base < cp.min) throw bad('Coupon requires a minimum amount of Rs ' + cp.min);
    const raw = cp.type === 'pct' ? Math.round((c.base - discount) * cp.val / 100) : cp.val;
    discount += Math.min(raw, cp.max || raw);
    couponInfo = cp.code;
  }
  const taxable = Math.max(0, c.base - discount);
  const gst = Math.round(taxable * (p.gstPct || 0) / 100);
  return {
    family: c.family, included: c.included, label: c.label,
    base: c.base, discount, gst, final: taxable + gst,
    currency: p.currency, coupon: couponInfo, gstPct: p.gstPct || 0
  };
}

/* Idempotency: returns the previously stored result for a key within its scope. */
function idemGet(key, scope) {
  if (!key) return null;
  const r = db.prepare('SELECT result FROM idempotency_keys WHERE key=? AND scope=?').get(String(key).slice(0, 120), scope);
  return r ? JSON.parse(r.result) : null;
}
function idemPut(key, scope, result) {
  if (!key) return;
  db.prepare('INSERT OR IGNORE INTO idempotency_keys(key,scope,result,created_at) VALUES(?,?,?,?)')
    .run(String(key).slice(0, 120), scope, JSON.stringify(result), Date.now());
}

const BILLING_STATES = ['FREE', 'PENDING_PAYMENT', 'PAID', 'REFUNDED', 'CANCELLED'];

module.exports = { pricing, includedCount, usedCount, classify, quoteFor, idemGet, idemPut, BILLING_STATES, DEFAULTS };

File: backend-python/app/services/kundali_billing.py
"""Kundali commercial model — port of server/services/kundaliBilling.js.

One customer account includes a plan-wise number of free (personal) kundalis;
every additional — and every FAMILY MEMBER — kundali is a chargeable request.
All prices, GST, discounts, coupon eligibility and free counts come from the
`kundali_pricing` setting (Admin > Kundali Management > Pricing). Nothing is
hard-coded here beyond safe defaults for a fresh database.

Billing states: FREE | PENDING_PAYMENT | PAID | REFUNDED | CANCELLED"""
import json

from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from ..models import Coupon, Kundali, Setting, User
from ..util import bad

DEFAULTS = {
    "active": True,
    "currency": "INR",
    "personalPrice": 0,        # paid only after the plan's free quota is used up
    "familyPrice": 499,        # every family-member kundali is chargeable
    "additionalPrice": 499,    # extra personal kundalis beyond the quota
    "gstPct": 5,
    "discountPct": 0,
    "couponEligible": True,
    "freeCounts": {"customer": 1, "plus": 2, "premium": 5},
}


async def _get_setting_raw(db: AsyncSession, key: str):
    row = (await db.execute(select(Setting).where(Setting.key == key))).scalar_one_or_none()
    if not row or not row.value:
        return None
    try:
        return json.loads(row.value)
    except (ValueError, TypeError):
        return None


async def pricing(db: AsyncSession) -> dict:
    stored = await _get_setting_raw(db, "kundali_pricing")
    p = dict(DEFAULTS)
    if isinstance(stored, dict):
        p.update(stored)
    return p


async def included_count(db: AsyncSession, user: User) -> int:
    p = await pricing(db)
    if getattr(user, "premium", 0):
        return p["freeCounts"]["premium"]
    if user.plus:
        return p["freeCounts"]["plus"]
    return p["freeCounts"]["customer"]


# Personal kundalis used against the plan quota (family kundalis never consume it).
async def used_count(db: AsyncSession, user_id: str) -> int:
    n = (await db.execute(
        select(func.count()).select_from(Kundali).where(
            Kundali.customer_id == user_id,
            func.coalesce(Kundali.relationship, "") == "",
        ))).scalar_one()
    return int(n)


# Decide billing for a request. `relationship` empty => personal kundali.
async def classify(db: AsyncSession, user: User, relationship: str | None) -> dict:
    p = await pricing(db)
    if not p["active"]:
        raise bad("Kundali generation is currently unavailable")
    rel = str(relationship or "").strip()
    if rel:
        return {"family": True, "included": False, "base": p["familyPrice"], "label": rel}
    used = await used_count(db, user.id)
    inc = await included_count(db, user)
    if used < inc:
        return {"family": False, "included": True, "base": 0, "label": "Included"}
    return {"family": False, "included": False, "base": p["additionalPrice"], "label": "Additional"}


# Full quote: base -> plan/special discount -> coupon -> GST. Amounts in rupees (int).
async def quote_for(db: AsyncSession, user: User, relationship: str | None = None,
                    coupon: str | None = None) -> dict:
    p = await pricing(db)
    c = await classify(db, user, relationship)
    discount = round(c["base"] * (p.get("discountPct") or 0) / 100)
    coupon_info = None
    if coupon and c["base"] > 0:
        if not p.get("couponEligible"):
            raise bad("Coupons do not apply to kundali purchases")
        cp = (await db.execute(select(Coupon).where(
            Coupon.code == str(coupon).upper(), Coupon.active == 1))).scalar_one_or_none()
        if not cp:
            raise bad("Coupon code is not valid")
        if cp.min and c["base"] < cp.min:
            raise bad("Coupon requires a minimum amount of Rs " + str(cp.min))
        raw = round((c["base"] - discount) * cp.val / 100) if cp.type == "pct" else cp.val
        discount += min(raw, cp.max if cp.max else raw)
        coupon_info = cp.code
    taxable = max(0, c["base"] - discount)
    gst = round(taxable * (p.get("gstPct") or 0) / 100)
    return {
        "family": c["family"], "included": c["included"], "label": c["label"],
        "base": c["base"], "discount": discount, "gst": gst, "final": taxable + gst,
        "currency": p["currency"], "coupon": coupon_info, "gstPct": p.get("gstPct") or 0,
    }


# Idempotency: returns the previously stored result for a key within its scope.
async def idem_get(db: AsyncSession, key: str | None, scope: str):
    if not key:
        return None
    from ..models import IdempotencyKey
    r = (await db.execute(select(IdempotencyKey).where(
        IdempotencyKey.key == str(key)[:120], IdempotencyKey.scope == scope))).scalar_one_or_none()
    try:
        return json.loads(r.result) if r and r.result else None
    except (ValueError, TypeError):
        return None


async def idem_put(db: AsyncSession, key: str | None, scope: str, result) -> None:
    if not key:
        return
    from ..models import IdempotencyKey
    exists = (await db.execute(select(IdempotencyKey).where(
        IdempotencyKey.key == str(key)[:120], IdempotencyKey.scope == scope))).scalar_one_or_none()
    if exists:
        return
    db.add(IdempotencyKey(key=str(key)[:120], scope=scope,
                          result=json.dumps(result), created_at=int(now_ms())))


def now_ms() -> int:
    import time
    return int(time.time() * 1000)


BILLING_STATES = ['FREE', 'PENDING_PAYMENT', 'PAID', 'REFUNDED', 'CANCELLED']
